core/domain/suggestion_services.py

```diff
@@ -54,18 +75,68 @@
 # suggestion opportunities. For instance, for translation suggestions the
 # emphasized text is the translation. Similarly, for question suggestions the
 # emphasized text is the question being asked.
-SUGGESTION_EMPHASIZED_TEXT_GETTER_FUNCTIONS = {
+SUGGESTION_EMPHASIZED_TEXT_GETTER_FUNCTIONS: Dict[
+    str, Callable[[suggestion_registry.BaseSuggestion], str]
+] = {
     feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT: (
         lambda suggestion: suggestion.change.translation_html),
+    # Here, change is of type BaseChange and all attributes on BaseChange are
+    # created dynamically except cmd, and the type of all dynamically created
+    # attributes are considered as string (str) type. So, question_dict
+    # is also considered as string type but here we are using it as a Dict
+    # type which causes MyPy throws an error. Thus to avoid the error, we
+    # used ignore here.
```

Comment:
Hmm, I remember we have this issue with `BaseChange` quite a lot, do you think there is any possible change that would strengthen the typing?

Reply:
I think looking at the current scenario we are doing the best solution ( as far I know ):
because in base change every attribute is defined dynamically and most of the attributes have str type that's why the __getattr__ method has str return type. 
Also, we are allowing only a limited number of types to enter in _init_ method of BaseChange. 

Now, I think the only concerned left is that can we add different types to different attributes ( like: question_dict should have QuestionDict without any explicit declaration ), for this i think we can do the following:
```python 
class BaseChange:
    """Domain object for changes made to storage models' domain objects."""
    if MYPY:
        question_dict: QuestionDict
        new_value: PossibleNewValueTypes
        old_value: PossibleOldValueTypes
        .
        .
        <<other attributes>>
```
in this way we have a sub-set of types for each individual attribute. The problem with this is that it only reduces 5 to 10% percent of assertions ( I'm not sure ) because if new_value has a subset of 3 types ( str, int, bool) then we have to use assertion again to narrow down it to a specific type. But for cases like question_dict we don't have to use assertion anymore. Also, This can be implemented once we have enough info about each attribute may be after covering the domain folder ( at least ).

@vojtechjelinek PTAL! and let me know about this approach.

Reply:
Yeah, that looks good.